### PR TITLE
[BED-322] Void lookups support

### PR DESCRIPTION
**Pull request by csucu**

### CODE REVIEW

File: resolver_miekg.go [hunk 1]
```diff
@@ -168,100 +169,130 @@ func (r *miekgDNSResolver) exchange(req *dns.Msg) (*dns.Msg, error) {
...
+		minTTL = 0
+	}
+
+	if len(res.Answer) == 0 && res.Rcode == dns.RcodeSuccess {
+		void = true
+	}
+
+	extras := &ResponseExtras{
+		TTL:  time.Duration(minTTL) * time.Second,
+		Void: void,
```
Comment (dmotylev) [inline resolver_miekg.go]:
```suggestion
	extras := &ResponseExtras{
		TTL:  time.Duration(minTTL) * time.Second,
		Void: len(res.Answer) == 0 && res.Rcode == dns.RcodeSuccess,
```

File: resolver_miekg.go [hunk 2]
```diff
@@ -168,100 +169,130 @@ func (r *miekgDNSResolver) exchange(req *dns.Msg) (*dns.Msg, error) {
...
+	}
+
 	if as == 0 {
-		ttl = 0
+		minTTL = 0
+	}
+
+	extras := &ResponseExtras{
+		TTL:  time.Duration(minTTL) * time.Second,
+		Void: void,
```
Comment (dmotylev) [inline resolver_miekg.go]: inline `void` expression

File: resolver_miekg.go [hunk 3]
```diff
@@ -386,41 +419,51 @@ func (r *miekgDNSResolver) MatchMX(name string, matcher IPMatcherFunc) (bool, ti
...
+	}
+
 	if len(ptrs) == 0 {
-		ttl = 0
+		minTTL = 0
+	}
+
+	extras := &ResponseExtras{
+		TTL:  time.Duration(minTTL) * time.Second,
+		Void: void,
```
Comment (dmotylev) [inline resolver_miekg.go]: inline `void` expression

**Review by dmotylev — CHANGES_REQUESTED** [review]
Functionally it seems good; however, we need the following:
- add tests to cover added functionality;
- follow documentation style and explain what is void lookup and include references to RFC

File: resolver_limited.go [hunk 4]
```diff
@@ -91,12 +121,19 @@ func (r *LimitedResolver) MatchMX(name string, matcher IPMatcherFunc) (bool, *Re
...
 	limit := int32(r.mxQueriesLimit)
-	return r.resolver.MatchMX(name, func(ip net.IP, name string) (bool, error) {
+	found, extras, err := r.resolver.MatchMX(name, func(ip net.IP, name string) (bool, error) {
 		if atomic.AddInt32(&limit, -1) < 1 {
 			return false, ErrDNSLimitExceeded
 		}
 		return matcher(ip, name)
 	})
+	if extras != nil && extras.Void {
+		if !r.canPerformVoidLookup() {
```
Comment (dmotylev) [inline resolver_limited.go]: Did you consider adding 
```golang
func isVoid(v *ResponseExtrase) bool { return v!=nil && v.Void }
```
?

then this bit could be simplified to 
```golang
if isVoid(extras) && !r.canPerformVoidLookup() { ... }
```

File: resolver_limited_test.go [hunk 5]
```diff
@@ -87,10 +87,22 @@ func TestLimitedResolver(t *testing.T) {
...
+		dns.HandleFunc("void.test.", Zone(map[uint16][]string{}))
+		defer dns.HandleRemove("void.test.")
+
+		r := NewLimitedResolver(testResolver, 6, 6, 2)
+		_, _, _ = r.LookupTXTStrict("void.test.")
+		_, _, _ = r.LookupTXTStrict("void.test.")
+		_, _, err := r.LookupTXTStrict("void.test.")
+		if err != ErrDNSVoidLookupLimitExceeded {
+			t.Errorf("LookupTXTStrict got: %v; want ErrDNSVoidLookupLimitExceeded", err)
+		}
```
Comment (dmotylev) [inline resolver_limited_test.go]: It does not check if resolver allowed to do first 2 lookups

**Review by dmotylev — CHANGES_REQUESTED** [review]
Overall good; need a fix for the void lookup test and consider minor code refactoring

File: spf.go [hunk 6]
```diff
@@ -65,32 +66,51 @@ func newMissingMacrosError(domain string, macros []string) error {
...
+type ResponseExtras struct {
+	TTL  time.Duration // Minimum TTL of the DNS response
+	Void bool          // Indicates if the response is a result of a DNS void lookup.
+
+	// A DNS void lookup, as defined in Section 4.6.4 of RFC 7208 (https://datatracker.ietf.org/doc/html/rfc7208#section-4.6.4),
+	// is a query for a domain that is intentionally configured to have no associated DNS records,
+	// such as an explicit configuration for a "blackhole" or an intentionally nonexistent domain.
+	// This type of query typically returns a response with no relevant DNS records (e.g., NXDOMAIN),
+	// and the 'Void' field in this struct is set to 'true' to indicate that the response resulted from such a lookup.
+}
```
Comment (dmotylev) [inline spf.go]: You pass ResponseExtras as a reference everywhere to use a nil value when some non-DNS error occurs. This is ok, but it forces users to do additional checks for nil. It increases pollution. Making fields private and using methods that will consider nil receivers would be beneficial. Like
```golang
func (x *ResponseExtras) TTL() time.Duration {
  if x==nil { return 0 }
  return x.ttl
}

func (x *ResponseExtras) IsVoid() bool {
   return x!=nil && x.void
}
```

File: resolver_limited.go (hunk 4 above)
Comment (csucu) [reply resolver_limited.go]: done

File: resolver_limited_test.go (hunk 5 above)
Comment (csucu) [reply resolver_limited_test.go]: done

File: spf.go (hunk 6 above)
Comment (csucu) [reply spf.go]: done

File: spf.go [hunk 7]
```diff
@@ -113,6 +113,17 @@ type Resolver interface {
...
+	if x == nil {
+		return 0
+	}
+	return x.TTL
+}
+
+func (x *ResponseExtras) IsVoid() bool {
+	return x != nil && x.Void
+}
+
```
Comment (dmotylev) [inline spf.go]: Nope, having two different access models worsen the design.
We either have to use public fields (`TTL` and `Void`) and no getters or private fields (`ttl` and `void`) and getters (`TTL()` and `Void()`).

In the latter, we also need the constructor.

Also, it is better to place methods next to the structure

Comment (csucu) [reply spf.go]: done

File: parser.go [hunk 8]
```diff
@@ -730,7 +757,11 @@ func (p *parser) handleExplanation(t *token) (string, error) {
 		return "", SyntaxError{t, newInvalidDomainError(domain)}
 	}
 
-	txts, _, err := p.resolver.LookupTXT(NormalizeFQDN(domain))
+	txts, extras, err := p.resolver.LookupTXT(NormalizeFQDN(domain))
+	if extras.Void() {
+		p.fireVoidLookup()
+	}
+
```
Comment (dmotylev) [inline parser.go]: We did not have previously fired the Explanation event to the listener. Fire void lookup without it will mess up the walker state

**Review by dmotylev — CHANGES_REQUESTED** [review]
We should not send a "void lookup" event just for having it. We must ensure the listener can attach this signal to the correct directive.

There are at least two design choices: 

- extend current `fireXXX` methods with additional info
- or make sure we `fireVoidLookup()` when the listener knows to what directive attach it; saying so we can not send it before or without corresponding `fireXXX` call

File: parser.go (hunk 8 above)
Comment (csucu) [reply parser.go]: updated

**Review by dmotylev — CHANGES_REQUESTED** [review]
Commit [414b165](https://github.com/redsift/spf/pull/20/commits/414b165bb541193bb0bd947ed3faf482635101c0) is against the design of the Listener.

We should not expose implementation details to the package users. And we pass all needed info as parameters.

**Review by dmotylev — APPROVED** [review]

**Pull request merged**